Add BidirectionalCollection.trimming

karwa opened the pull request:
<!--
    Thanks for contributing to Swift Algorithms!

    Before you submit your request, please replace each paragraph
    below with the relevant details, and complete the steps in the
    checklist by placing an 'x' in each box:
    
    - [x] I've completed this task
    - [ ] This task isn't completed
-->

### Description
Adds a '.trimming' method to BidirectionalCollection, returning a slice with matching elements removed from the start and end.

### Detailed Design

```swift
extension BidirectionalCollection {

  /// Returns a `SubSequence` formed by discarding all elements at the start and end of the Collection
  /// which satisfy the given predicate.
  ///
  /// e.g. `[2, 10, 11, 15, 20, 21, 100].trimmimg(where: { $0.isMultiple(of: 2) })` == `[11, 15, 20, 21]`
  ///
  /// - parameters:
  ///    - predicate:  A closure which determines if the element should be omitted from the resulting slice.
  ///
  @inlinable
  public func trimming(where predicate: (Element) throws -> Bool) rethrows -> SubSequence
}
```

### Documentation Plan
No existing guides need updating AFAIK. I haven't written a guide for this feature: is it necessary for every API to include this parallel documentation?

### Test Plan
Tests are included.

### Source Impact
Additive.

### Checklist
- [X] I've added at least one test that validates that my change is working, if appropriate
- [?] I've followed the code style of the rest of the project
- [X] I've read the [Contribution Guidelines](CONTRIBUTING.md)
- [X] I've updated the documentation if necessary

karwa commented:
In terms of code style, I think my defaults match this project (except that I use a 120 line length).

Does this project just use the default swift-format settings?

timvermeulen commented on Sources/Algorithms/Trim.swift:
```diff
@@ -0,0 +1,40 @@
+    var sliceEnd = endIndex
+    // Consume elements from the front.
+    while sliceStart != sliceEnd, try predicate(self[sliceStart]) {
+      sliceStart = index(after: sliceStart)
```
```suggestion
      formIndex(after: &sliceStart)
```

natecook1000 commented on Sources/Algorithms/Trim.swift:
```diff
@@ -0,0 +1,40 @@
+    var sliceStart = startIndex
+    var sliceEnd = endIndex
+    // Consume elements from the front.
+    while sliceStart != sliceEnd, try predicate(self[sliceStart]) {
```
Can you use existing stdlib methods here, instead of searching for the start and end manually?

karwa replied:
I can use `firstIndex(where:)` to find the start, but not `lastIndex(where:)` to find the end (since we actually want the index _after_ the last index to match the predicate; this would mean an extra call to `index(after:)`).

I've tried to improve the documentation, too. It might be worth adding a `String.trimmed()` function which defaults to trimming whitespace.

karwa commented:
@natecook1000 I've added a guide document. It may be a little verbose, I don't know 😅. That's just my style - it's how I write, but I did my best.

I agree with `trim` being a better root verb -- or `trimmed` as an adjective, which reads like "this is the collection, trimmed of these elements", with the verb form perhaps replacing/expanding on the mutating `removeFirst/Last(Int)` methods. I've mentioned the various methods in the standard library in the guide document. I tried not to get too bogged-down in that discussion, otherwise it'd turn more in to a swift-evolution pitch than a useful document for people using this library.

karwa commented:
@natecook1000 is there anything remaining for me to do here?

natecook1000 commented on Guides/Trim.md:
```diff
@@ -0,0 +1,116 @@
+
+This method requires `BidirectionalCollection` for an efficient implementation which visits as few elements as possible.
+
+A less-efficient implementation is _possible_ for any `Collection`, which would involve always traversing the
+entire collection. This implementation is not provided, as it would mean developers of generic algorithms who forget
+to add the `BidirectionalCollection` constraint will recieve that inefficient implementation:
```
Agreed, `BidirectionalCollection` is the right place for this.

xwu replied:
```suggestion
to add the `BidirectionalCollection` constraint will receive that inefficient implementation:
```

natecook1000 commented on Sources/Algorithms/Trim.swift:
```diff
@@ -0,0 +1,48 @@
+  /// Returns a `SubSequence` formed by discarding all elements at the start and end of the Collection
+  /// which satisfy the given predicate.
+  ///
+  /// Trimming is a common operation in computing, often referring to removing leading and trailing whitespace
+  /// from a string[1].
```
I don't think we need this explanation. Let's introduce the example, instead — something like:

> This example uses `trimmed(where:)` to get a substring without the white space at the beginning and end of the string.

natecook1000 commented on Sources/Algorithms/Trim.swift:
````diff
@@ -0,0 +1,48 @@
+  ///  print(myString.trimmed(where: \.isWhitespace)) // "hello, world"
+  ///  ```
+  ///
+  /// [1]: https://en.wikipedia.org/wiki/Trimming_(computer_programming)
````
```suggestion
```

We should mention that if the entire collection is trimmed, the  result is an empty subsequence at the end of the collection.

karwa replied:
Personally I'd prefer to leave the position of the empty slice unspecified. If we decided we wanted to change this to first trim elements from the tail and then the head, the result when trimming the entire collection would be an empty subsequence at `startIndex`.

natecook1000 reviewed (commented):
Looking good! Just a couple notes about the documentation.

xwu commented on Guides/Trim.md:
```diff
@@ -0,0 +1,116 @@
+
+### Complexity
+
+Calling this methods is O(_n_).
```
```suggestion
Calling this method is O(_n_).
```

xwu commented on Sources/Algorithms/Trim.swift:
```diff
@@ -0,0 +1,48 @@
+    while sliceStart != sliceEnd {
+      let idxBeforeSliceEnd = index(before: sliceEnd)
+      guard try predicate(self[idxBeforeSliceEnd]) else {
+        return self[sliceStart..<sliceEnd]
```
Q: can these bounds be safely unchecked too?

karwa replied:
I'm not 100% sure. Since `sliceEnd` is formed by walking backwards from `endIndex` and has not yet tested equal to `sliceStart`, the only way this bounds check will fail is if either your collection or the index's `Comparable` conformance is broken.

For the empty result, we already checked that `sliceStart == sliceEnd` as part of the `while` loop condition, so it is safe to omit the bounds check. A broken conformance would never get to that part.

xwu commented on Sources/Algorithms/Trim.swift:
```diff
@@ -0,0 +1,48 @@
+
+extension BidirectionalCollection {
+
+  /// Returns a `SubSequence` formed by discarding all elements at the start and end of the Collection
```
Nit: I _think_ this project's style is to wrap to 80 columns.

karwa replied:
I ran this through swift-format (the project doesn't have a format file, so I assume it's the default settings?), but it doesn't seem to wrap comments. Will fix.

xwu commented on Guides/Trim.md:
````diff
@@ -0,0 +1,116 @@
+which satisfy the given predicate.
+
+```swift
+let results = [2, 10, 11, 15, 20, 21, 100].trimmed(where: { $0.isMultiple(of: 2) })
````
Just a teeny bit of bikeshedding. Because trailing closure syntax causes `where:` to be omitted, there can be a little bit of ambiguity as to whether the closure determine's what's included or what's excluded. It so happens that, in English, "trimming" can take a direct object, while "trimmed" cannot, such that `[1, 2, 3].trimming { $0 == 1 }` suggests that every element equal to one is being dropped a little more strongly than "trimmed" would.

karwa replied:
Yeah, that sounds good. I like `trimmed`.

xwu commented on Guides/Trim.md:
```diff
@@ -0,0 +1,116 @@
+[[Source](https://github.com/apple/swift-algorithms/blob/main/Sources/Algorithms/Trim.swift) |
+ [Tests](https://github.com/apple/swift-algorithms/blob/main/Tests/SwiftAlgorithmsTests/TrimTests.swift)]
+
+Returns a `SubSequence` formed by discarding all elements at the start and end of the Collection
```
I think the prevailing style is to use code voice to refer to a `Thing` or otherwise to refer to the thing being modeled in lowercase. A `Collection` is a protocol for collections; a `String` models a string.

```suggestion
Returns a `SubSequence` formed by discarding all elements at the start and end of the collection
```

xwu commented on Guides/Trim.md:
````diff
@@ -0,0 +1,116 @@
+print(myString.trimmed(where: \.isWhitespace)) // "hello, world"
+```
+
+Trimming is a common operation in computing. It is particularly for String processing, but also useful in
````
```suggestion
Trimming is a common operation in computing. It is particularly for string processing, but also useful in
```

xwu commented on Sources/Algorithms/Trim.swift:
```diff
@@ -0,0 +1,48 @@
+
+extension BidirectionalCollection {
+
+  /// Returns a `SubSequence` formed by discarding all elements at the start and end of the Collection
```
```suggestion
  /// Returns a `SubSequence` formed by discarding all elements at the start and end of the collection
```

xwu reviewed (commented):
A remark on documentation.

karwa commented:
Apologies for the delay. I've addressed your comments @xwu and @natecook1000 - documentation has been improved, and I think Xiaodi makes a good case for the name `trimming`. Code in the tests reads quite nicely after this, e.g.:

```
let results = [2, 10, 11, 15, 20, 21, 100].trimming { $0.isMultiple(of: 2) }
```
